cli/src/commands/rebase.rs

```diff
@@ -873,17 +874,33 @@ fn move_commits(
             if let Some(new_child_parents) = new_children_parents.get(commit_id) {
                 new_child_parents.clone()
             }
-            // Commits in the target set should persist only rebased parents from the target
-            // sets.
-            else if let Some(target_commit_parents) =
-                target_commits_internal_parents.get(commit_id)
-            {
-                // If the commit does not have any parents in the target set, it is one of the
-                // commits in the root set, and should be rebased onto the new destination.
-                if target_commit_parents.is_empty() {
+            // Commit is in the target set.
+            else if target_commit_ids.contains(commit_id) {
+                // If the commit is a root of the target set, it should be rebased onto the new destination.
+                if target_roots.contains(commit_id) {
                     new_parent_ids.clone()
-                } else {
-                    target_commit_parents.clone()
+                }
+                // Otherwise:
+                // 1. Keep parents which are within the target set.
+                // 2. Replace parents which are outside the target set but are part of the
+                //    connected target set with their ancestor commits which are in the target
+                //    set.
+                // 3. Keep other parents outside the target set if they are not descendants of the
+                //    new children of the target set.
+                else {
+                    let mut new_parents = vec![];
```

Comment:
Perhaps, `new_parents` can be precomputed in a way that skipped ancestor edges are also preserved?

```
    create_commit(&test_env, &repo_path, "a", &[]);
    create_commit(&test_env, &repo_path, "b", &["a"]);
    create_commit(&test_env, &repo_path, "c", &["a"]);
    create_commit(&test_env, &repo_path, "d", &["a"]);
    create_commit(&test_env, &repo_path, "e", &["c", "b"]);
    create_commit(&test_env, &repo_path, "f", &["e"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @  f
    ◉    e
    ├─╮
    │ ◉  b
    ◉ │  c
    ├─╯
    │ ◉  d
    ├─╯
    ◉  a
    ◉
    "###);

    // TODO: edge f->e->b is lost
    test_env.jj_cmd_ok(&repo_path, &["rebase", "-r", "c|f", "-d", "d"]);
    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
    @  f
    ◉  c
    ◉  d
    │ ◉  e
    ╭─┤
    │ ◉  b
    ├─╯
    ◉  a
    ◉
    "###);
```

Reply:
Interesting case. I don't think I would actually expect the dependency on `b` to be preserved.

Reply:
My thought process for the implementation is that when selecting the target commits to rebase, we only want to preserve ancestry information for those commits, not any other commits. However, I'm happy to change the implementation if a consensus is reached. My personal use of this feature has largely been limited to rebasing linear stacks of commits to before/after other locations, so maybe others with different usecases might want to make suggestions as well.

Reply:
Hmm, in my mental model, `rebase -r c|f` is roughly equivalent to `rebase -r c::f` and `abandon e'`. So I expect the `f->b` edge is preserved so long as `rebase -r c::f` works that way.

Reply:
Just commenting why I think the `f->e->b` edge should be preserved:
1. disconnected `-r` targets are rebased as if they were all connected
2. in order to achieve that, internal edges across skipped nodes are preserved
3. so it seems better to apply the same rule to external edges

In practice, I wouldn't try this kind of weird rebase, and I wouldn't mind if `f->e->b` was removed. My point is that it's easier to reason about the behavior if we have a consistent model (such as `rebase -r c|f` is equivalent to `rebase --keep -r c::f` + `abandon c|f` + `abandon e'`.)

Reply:
I guess making the logic more consistent is also a good idea. I will update the PR to change this.

Reply:
I think the other reasonable behavior is to drop this PR and let `-s x` and `-r x::` be different. I think I would actually prefer that. That seems even more consistent. Then we don't even have to worry about treating the roots of the connected set differently. It's weird to me if `rebase -r c|f -d a` preserves the edge but `rebase -r f -d a` does not. It doesn't seem that `f` should depend on `b` any more because we're also rebasing `c`. Thoughts?

Reply:
Yet another option is to make `-s` not preserve external edges. I copied it from Mercurial but I'm not sure how useful that behavior is.

Reply:
> I think the other reasonable behavior is to drop this PR and let `-s x` and `-r x::` be different.

That sounds also good to me.

> It's weird to me if `rebase -r c|f -d a` preserves the edge but `rebase -r f -d a` does not. It doesn't seem that `f` should depend on `b` any more because we're also rebasing `c`.

Since `rebase -r c|f -d a` is not identical to `rebase -r c -d a && rebase -r f -d a`, I won't be surprised if `-r c|f` preserve the `f->e->b` edge, but `-r f` doesn't.

> Yet another option is to make `-s` not preserve external edges. I copied it from Mercurial but I'm not sure how useful that behavior is.

I occasionally do `rebase -s` one branch whose descendants are merged with the other branches. In this case, the current `rebase -s` behavior is convenient.

Reply:
Based on this discussion, should I close this PR? I can open a separate one which moves the `move_commits` function into the `rewrite` module.

Also, is https://github.com/martinvonz/jj/pull/3607 still desired? (I was working initially working on this PR to make the behavior of `rebase -r/-s` consistent before migrating `rebase -s` to the `move_commits` function to add support for `--before/after`.)

Reply:
Yes, my inclination is to drop this PR and let `jj rebase -s foo` and `jj rebase -r foo::` behave differently. Sorry to have wasted your time :(

> Also, is #3607 still desired?

I think it's enough to support `-A/-B` with `-r`, and it might get complicated to support it with `-s` and its different behavior for merges into the set.

Reply:
Apart from `--insert-before/after`, I was thinking of `rebase --keep` as well. I was planning to migrate the existing `rebase -s/-b` to use `move_commits`, then from there modify `move_commits` to allow keeping the original target commits as well. WDYT about that? (I don't think think it will be very complicated to add a flag to persist external merge parents for `-s`/`-b`.)

Reply:
That all sounds good to me. Thanks!

Reply:
#3788 made me think about this thread again. Similar to `-s`, it's a bit ugly that we can't get the same behavior with `-r`. Maybe we should add a `--preserve-external-edges` to preserve edges coming into revisions that are not the roots of the set. I think we could then replace the current `-s foo` implementation by `-r foo:: --preserve-external-edges` then, and the current `-b foo -d bar` by `-r '(bar..foo)::' --preserve-external-edges`.

Reply:
Do we need `--preserve-external-edges` flag? iirc, there are some oddity in `--preserve-external-edges -rREVS` behavior (especially when `REVS` aren't contiguous), but would it matter in practice?

I think it's be a bit confusing if `-s`/`-b` implies `--preserve-external-edges`, but `-r` doesn't. If the "preserve" behavior is practically good, it can be turned on by default.